src/bsp/src/stm32/src/SpiEsp.cpp

```diff
@@ -0,0 +1,211 @@
+        }
+        // This will be sent on next header. Payload has priority over headers.
+        m_inboundMessage.m_sizeBytes = m_inboundHeader->txSizeWord << 2;
+        if (m_inboundMessage.m_sizeBytes == m_outboundHeader.rxSizeWord << 2 &&
+            m_inboundMessage.m_sizeBytes != 0) {
+            rxLengthBytes = m_inboundHeader->txSizeWord << 2;
+            m_outboundHeader.systemState.stmSystemState.failedCrc = 0;
+            m_rxState = receiveState::RECEIVING_PAYLOAD;
+        } else {
+            rxLengthBytes = EspHeader::sizeBytes;
+            m_rxState = receiveState::RECEIVING_HEADER;
+        }
+        break;
+    case receiveState::RECEIVING_PAYLOAD:
+        m_outboundHeader.systemState.stmSystemState.failedCrc = 0;
+        rxLengthBytes = m_inboundHeader->txSizeWord << 2;
+        break;
+    case receiveState::VALIDATE_CRC:
+        if (m_crc.calculateCRC32(m_inboundMessage.m_data.data(),
+                                 m_inboundMessage.m_sizeBytes - CRC32_SIZE) !=
+            *(uint32_t*)&m_inboundMessage.m_data[m_inboundMessage.m_sizeBytes - CRC32_SIZE]) {
+            m_outboundHeader.systemState.stmSystemState.failedCrc = 1;
+        } else {
+            m_logger.log(LogLevel::Info, "ESP says: %s", m_inboundMessage.m_data.data());
+        }
+        m_inboundMessage.m_sizeBytes = 0;
+        m_rxState = receiveState::RECEIVING_HEADER;
+        rxLengthBytes = EspHeader::sizeBytes;
+        break;
+    case receiveState::ERROR:
+        rxLengthBytes = EspHeader::sizeBytes;
+        m_rxState = receiveState::RECEIVING_HEADER;
+        break;
+    }
+    if (m_inboundRequest && m_txState == transmitState::IDLE) {
+        m_txState = transmitState::SENDING_HEADER;
+    }
+
+    // Transmitting state machine
+    switch (m_txState) {
+    case transmitState::IDLE:
+        break;
+    case transmitState::SENDING_HEADER:
+        updateOutboundHeader();
+        txLengthBytes = EspHeader::sizeBytes;
+        txBuffer = (uint8_t*)&m_outboundHeader;
+        break;
+    case transmitState::SENDING_PAYLOAD:
+        txLengthBytes = m_outboundMessage.m_sizeBytes;
+        txBuffer = m_outboundMessage.m_data.data();
+        m_outboundMessage.m_sent = true;
+        break;
+    case transmitState::ERROR:
+        HAL_GPIO_WritePin(ESP_CS_GPIO_Port, ESP_CS_Pin, GPIO_PIN_SET);
+        break;
+    }
+
+    if ((m_inboundRequest || m_outboundMessage.m_sizeBytes != 0) &&
+        m_txState != transmitState::ERROR && m_rxState != receiveState::ERROR) {
+        HAL_GPIO_WritePin(ESP_CS_GPIO_Port, ESP_CS_Pin, GPIO_PIN_RESET);
+        vTaskDelay(20);
+        uint32_t finalSize = std::max(txLengthBytes, rxLengthBytes);
+        memset(m_inboundMessage.m_data.data(), 0, ESP_SPI_MAX_MESSAGE_LENGTH);
+        if (xSemaphoreTake(m_semaphore, (TickType_t)10) == pdTRUE) {
+            EspSpi_TransmitReceiveDma(txBuffer, m_inboundMessage.m_data.data(), finalSize,
```

Comment:
Do we really need a semaphore here? won't execute always be called in the same thread?